{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9301b27b",
   "metadata": {},
   "source": [
    "# Resources for help  @brown\n",
    "\n",
    "### COBRE CBHD Computational Biology Core\n",
    "    - Office hours\n",
    "    - https://cbc.brown.edu\n",
    "    - slack channel on ccv-share\n",
    "    - cbc-help@brown.edu\n",
    "    \n",
    "### Center for Computation and Visualization\n",
    "    - Office hours\n",
    "    - https://ccv.brown.edu\n",
    "    - ccv-share slack channel\n",
    "    - https://docs.ccv.brown.edu\n",
    "    - support@ccv.brown.edu\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6ef1026-b070-47c9-bb91-08afb9447a9a",
   "metadata": {},
   "source": [
    "# How does Bioinformatics Data analysis proceed\n",
    "\n",
    "* Usually start with some workflows to process/pre-process Raw data \n",
    "* Once the pre-processing/processing is done then move to the data analysis stage, again usually some kind of statistical analysis\n",
    "* Finally, data might be visualized Summarized \n",
    "\n",
    "![typical workflow](imgs/bio_workflow.png)\n",
    "\n",
    "All the above steps require some form of external data, usually annotations and there are some fundamental data formats that we tend to work with for NGS analysis. There is a few standard terms/Jargon that will help you navigate the different file formats.\n",
    "\n",
    "ONE important aspect to remember is that this is all usually sequence related, meaning somehow ends up being localized to some part of the genome. \n",
    "\n",
    "# PRIMARY DATA FORMATS\n",
    "There are a few primary data formats that we will talk about \n",
    "- Fasta, Fastq\n",
    "- SAM,BAM, CRAM\n",
    "- GTF, GFF3, BED\n",
    "- VCF\n",
    "\n",
    "Here are some resources that we have used to create this workshop\n",
    "- https://rnnh.github.io/bioinfo-notebook/docs/file_formats.html#cram\n",
    "- https://bioinformatics.uconn.edu/resources-and-events/tutorials-2/file-formats-tutorial/#\n",
    "\n",
    "But be careful \n",
    "- https://www.hadriengourle.com/tutorials/file_formats/\n",
    "\n",
    "Obviously, there are tons of resources out in the WWW but there are at least 3 that we recommend and constantly use\n",
    "- Wikipedia: https://en.wikipedia.org/wiki/FASTA_format\n",
    "- UCSC: http://genome.ucsc.edu/FAQ/FAQformat.html#format6\n",
    "- To troubleshoot: https://www.biostars.org/  https://www.seqanswers.com/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26a5506c-efcc-435a-8819-951381e20d62",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Fasta format\n",
    "\n",
    "FASTA is a ubiquitous text-based format for representing nucleotide sequences or amino acid sequences. A FASTA file can contain one sequence or multiple sequences. If a FASTA file contains multiple sequences, it may sometimes be referred to as a “multi-FASTA” file.\n",
    "\n",
    "Each FASTA entry begins with a > (greater-than) symbol, followed by a comment on the same line describing the sequence that will follow. The actual sequence begins on the line after this comment. Another > (greater-than) symbol denotes the beginning of another FASTA entry (comment describing the sequence + the sequence itself).\n",
    "\n",
    "EXAMPLE FASTA FILE\n",
    "This example FASTA file contains two linear nucleotide sequences.\n",
    "```\n",
    ">gi|1817694395|ref|NZ_JAAGMU010000151.1| Streptomyces sp. SID7958 contig-52000002, whole genome shotgun sequence\n",
    "CCGGCTGGCGCGGCTGGCGCTGGCGGTGGGGCTGCGGCTGCTGGAGCTGGGGGTGGCGCTGGAGGCGCAC\n",
    "GGCCAGAACCTGCTGGTGGTGCTGTCGCCGTCCGGGGAGCCGCGGCGGCTGGTCTACCGCGATCTGGCGG\n",
    "ACATCCGGGTCTCCCCCGCGCGGCTGGCCCGGCACGGTATCCGGGTTCCGGACCTGCCGGCG\n",
    "\n",
    ">gi|1643051563|gb|SZWM01000399.1| Citrobacter sp. TBCS-14 contig3128, whole genome shotgun sequence\n",
    "GCACAGTGAGATCAGCATTCCGTTGGATCTACTGGTCAATCAAAACCTGACGCTGGGTACTGAATGGAAC\n",
    "CAGCAGCGCATGAAGGACATGCTGTCTAACTCGCAGACCTTTATGGGCGGTAATATTCCAGGCTACAGCA\n",
    "GCACCGATCGCAGCCCATATTCGAAAGCCGAGATCTTCTCTTTGTTTGCCGAAAACAACATG\n",
    "```\n",
    "\n",
    "note here that a multi-line fasta sequence is 80 characters or less (often 70). "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2d0fce3-d54a-42bf-8254-77acb36c9f36",
   "metadata": {
    "tags": []
   },
   "source": [
    "## So why are we concerned about FASTA\n",
    "\n",
    "Most (if not all) reference genomes are in FASTA format. \n",
    "\n",
    "There are many resources for downloading reference genome sequences, with some subtle differences between them. A few common ones are \n",
    "\n",
    "- UCSC: https://genome.ucsc.edu/cgi-bin/hgGateway\n",
    "- Ensembl: https://useast.ensembl.org/index.html\n",
    "- RefSeq: https://www.ncbi.nlm.nih.gov/refseq\n",
    "\n",
    "### Caveats\n",
    "\n",
    "There can be different naming conventions for chromosomes - \n",
    "- deflines from a UCSC human genome looks like this:\n",
    "```\n",
    ">chr1\n",
    ">chr2\n",
    ">chr3\n",
    ">chr4\n",
    ">chr5\n",
    "```\n",
    "\n",
    "Compared to Ensembl human genome deflines, which looks like this:\n",
    "\n",
    "```\n",
    ">1 dna:chromosome chromosome:GRCh38:1:1:248956422:1 REF\n",
    ">2 dna:chromosome chromosome:GRCh38:2:1:242193529:1 REF\n",
    ">3 dna:chromosome chromosome:GRCh38:3:1:198295559:1 REF\n",
    ">4 dna:chromosome chromosome:GRCh38:4:1:190214555:1 REF\n",
    ">5 dna:chromosome chromosome:GRCh38:5:1:181538259:1 REF\n",
    "```\n",
    "\n",
    "There are also differences in gene models – one example is PIK3CA (https://doi.org/10.1186/s12864-015-1308-8), where the Ensembl annotation is longer than the RefGene version.\n",
    "**Whichever reference genome you pick, be consistent (e.g., use a properly corresponding annotation file).**\n",
    "\n",
    "![Gene Models](imgs/gene_models.png)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a03bda07-baea-49d3-89c1-b149abf63a7d",
   "metadata": {},
   "source": [
    "## FASTA indices and dictionaries\n",
    "\n",
    "Many bioinformatics pipelines might want you to make a `.dict` dictionary for a FASTA file, especially if you are using something like PicardTools or the GATK pipeline\n",
    "- The output `.dict` file will just be a SAM header without any SAM records (we will talk about this later).\n",
    "\n",
    "Example `.dict`\n",
    "\n",
    "```\n",
    "@HD\tVN:1.0\tSO:unsorted\n",
    "@SQ\tSN:Ca22chr1A_C_albicans_SC5314\tLN:3188341\tM5:b1b7eebb81924852f04892f6cbe36cf4\tUR:file:///gpfs/data/cbc/references/Ene_SC5314_A22/SC5314_A22_chrA.fasta\n",
    "@SQ\tSN:Ca22chr2A_C_albicans_SC5314\tLN:2231883\tM5:8a8f4eda092437ebccc7974b759e14bd\tUR:file:///gpfs/data/cbc/references/Ene_SC5314_A22/SC5314_A22_chrA.fasta\n",
    "@SQ\tSN:Ca22chr3A_C_albicans_SC5314\tLN:1799298\tM5:29b2c5c3d01c692adaa77fd157421362\tUR:file:///gpfs/data/cbc/references/Ene_SC5314_A22/SC5314_A22_chrA.fasta\n",
    "@SQ\tSN:Ca22chr4A_C_albicans_SC5314\tLN:1603259\tM5:0de8b1a4e93ba5add01b2f21b50e8445\tUR:file:///gpfs/data/cbc/references/Ene_SC5314_A22/SC5314_A22_chrA.fasta\n",
    "@SQ\tSN:Ca22chr5A_C_albicans_SC5314\tLN:1190869\tM5:3218bf2cdb20033757d09cecc79a2d8d\tUR:file:///gpfs/data/cbc/references/Ene_SC5314_A22/SC5314_A22_chrA.fasta\n",
    "@SQ\tSN:Ca22chr6A_C_albicans_SC5314\tLN:1033292\tM5:7da684be586c16d8b1770d19335a770a\tUR:file:///gpfs/data/cbc/references/Ene_SC5314_A22/SC5314_A22_chrA.fasta\n",
    "@SQ\tSN:Ca22chr7A_C_albicans_SC5314\tLN:949580\tM5:0750e7c02925175a8aac6323b4373495\tUR:file:///gpfs/data/cbc/references/Ene_SC5314_A22/SC5314_A22_chrA.fasta\n",
    "@SQ\tSN:Ca22chrM_C_albicans_SC5314\tLN:40420\tM5:232094c1f5c073813a71486c621c2957\tUR:file:///gpfs/data/cbc/references/Ene_SC5314_A22/SC5314_A22_chrA.fasta\n",
    "@SQ\tSN:Ca22chrRA_C_albicans_SC5314\tLN:2286237\tM5:6f37f72edc9b1e5a64e17186676f737d\tUR:file:///gpfs/data/cbc/references/Ene_SC5314_A22/SC5314_A22_chrA.fasta\n",
    "```\n",
    "\n",
    "You might also be asked for `FASTA` or `FASTQ` indices or `.fai` files (http://www.htslib.org/doc/faidx.html).\n",
    "- These are tab-delimited text files and allow fast searching of the files.\n",
    "\n",
    "There are 5 tab-delimited columns for FASTA indices and 6 tab-delimited columns for FASTQ indices:\n",
    "- Name: Name of the reference sequence\n",
    "- Length: Total length of the reference sequence\n",
    "- Offset: Offset of the sequence’s first base (the position of the sequence in the fasta in terms of character count -- including newline characters, which means the newline encoding will matter here).\n",
    "- Linebases: Number of bases on each line\n",
    "- Linewidth: Number of bytes or characters in each line including the newline character.\n",
    "- Qualoffset: Offset of sequence’s first quality within the FASTQ file (FASTQ only).\n",
    "\n",
    "For example:\n",
    "```\n",
    "Ca22chr1A_C_albicans_SC5314\t3188341\t51\t60\t61\n",
    "Ca22chr2A_C_albicans_SC5314\t2231883\t3241583\t60\t61\n",
    "Ca22chr3A_C_albicans_SC5314\t1799298\t5510716\t60\t61\n",
    "Ca22chr4A_C_albicans_SC5314\t1603259\t7340054\t60\t61\n",
    "Ca22chr5A_C_albicans_SC5314\t1190869\t8970085\t60\t61\n",
    "Ca22chr6A_C_albicans_SC5314\t1033292\t10180853\t60\t61\n",
    "Ca22chr7A_C_albicans_SC5314\t949580\t11231417\t60\t61\n",
    "Ca22chrM_C_albicans_SC5314\t40420\t12196872\t60\t61\n",
    "Ca22chrRA_C_albicans_SC5314\t2286237\t12238017\t60\t61\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e497308e",
   "metadata": {},
   "source": [
    "# Useful Public Repositories\n",
    "\n",
    "## Reference data\n",
    "\n",
    "- Ensembl: https://useast.ensembl.org/index.html\n",
    "- UCSC Genome Browser : https://genome.ucsc.edu/cgi-bin/hgGateway\n",
    "- GenBank : https://www.ncbi.nlm.nih.gov/nucleotide/\n",
    "\n",
    "Lets explore a bit\n",
    " - Ensembl Biomart\n",
    " - UCSC/ Ensembl blat\n",
    " \n",
    "\n",
    "## Sequence Data\n",
    "\n",
    "Data in the SRA, European Bioinformatics Institute (EBI), and DNA Database of Japan (DDBJ) are all shared across all three resources.\n",
    "\n",
    "### NCBI Short Read Archive (SRA)\n",
    "\n",
    "The Sequence Read Archive (SRA) is the NIH’s repository of high-throughput sequencing data, including FASTQ files. \n",
    "\n",
    "The SRA is organized into four hierarchical levels: \n",
    "- Study (SRP/ERP/DRP) – overarching project\n",
    "- Sample (SRS/ERS/DRS) – biological samples (mouse or human cell cultures)\n",
    "- Experiment (SRX/ERX/DRX) – experiments on each sample (three treatment and three control samples)\n",
    "- Run (SRR/ERR/DRR) – sequencing run per experiment (might be multiple runs per experiment if it was split across multiple lanes).\n",
    "\n",
    " ![SRA structure](imgs/SRA_structure.png)\n",
    " \n",
    " Downloading the data in a usable format usually involves using the sra-toolkit https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=software\n",
    " \n",
    "### European Nucleotide Archive ENA\n",
    "\n",
    "You can also download reads from ENA using `wget` by searching for the Accession in the box labeled ‘Enter accession’ here: https://www.ebi.ac.uk/ena/browser/home \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b918e67a",
   "metadata": {},
   "source": [
    "# FASTQ\n",
    "The FASTQ format is an extension of FASTA that stores both biological sequences (usually nucleotide sequences) and their corresponding quality scores. Both the sequence letter and quality score are encoded with a single character for brevity.\n",
    "\n",
    "A FASTQ file normally uses four lines per sequence:\n",
    "\n",
    "- A line beginning with @ followed by a sequence identifier and optional description (like the comment line at in a FASTA file)\n",
    "- The raw sequence letters\n",
    "- A line beginning with +, sometimes followed by the same comment as the first line\n",
    "- A line encoding the quality values for the sequence in line 2, with the same numbers of symbols as letters in the sequence\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c2882e1-31d0-4125-86d4-6b702ac27e5a",
   "metadata": {},
   "source": [
    "EXAMPLE FASTQ FILE\n",
    "\n",
    "```\n",
    "@SRR8933535.1 1 length=75\n",
    "NAGGAAACAAAGGCTTACCCGTTATCATTTCCGCAAGAATGCACCCACACGACCATATATCAATGGATGTGGAGT\n",
    "+SRR8933535.1 1 length=75\n",
    "#AAAAEEEEEEEEEEEEEEEEEEEEEAEEEEEAEEEEEEEAEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEAE\n",
    "\n",
    "@SRR8933535.2 2 length=75\n",
    "NGAGGAGTGGTGGTAGTGTTGCTTGGTGGCAAAGATGTAGTTGGTGGGAAAGCTGAAGTGGTACCGTTGGTTGGA\n",
    "+SRR8933535.2 2 length=75\n",
    "#AAAAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEAAEEEE\n",
    "```\n",
    "\n",
    "Basically Fastq files are text files. \n",
    "\n",
    "### Why can't I open my fastq file in a text editor ?\n",
    "\n",
    "Most fastq files are compressed with Gzip so you will usually see the file extension `.gz` . This can result in significant space savings especially given that sequence data is highly repetitive compared to normal text. \n",
    "\n",
    "### How then do i generally view them?\n",
    "\n",
    "Use either `zcat < yourfastq.fq.gz | head -8` or `gunzip -c yourfastq.fq | head -8 ` to view the first two reads of your own FASTQ (you should probably just leave them all compressed as .gz files since they take up less space). \n",
    "\n",
    "\n",
    "### Single End vs Paired End Reads\n",
    "\n",
    "\n",
    "![Read types](imgs/pe_reads_img.png)\n",
    "\n",
    "- If your reads are single-end, each library will have just one read.\n",
    "- If your reads are paired-end, each library will (usually) have two reads.\n",
    "\n",
    "The paired-end read files should both have the same number of reads and be sorted in the same order based on their read names to maintain pair concordance.\n",
    "Be careful if trying to manually parse or filter your paired-end reads -- you want to maintain that pair concordance."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4462a93b-13e5-414a-a722-0e367dd6ed1a",
   "metadata": {},
   "source": [
    "### Where do I get some FASTQ Reads ?\n",
    "\n",
    "Some common sources are \n",
    "- SRA: The NCBI Short Read Archive https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=DRR059931\n",
    "- ENA: The European Nucleotide Archive https://www.ebi.ac.uk/ena/browser/view/PRJEB402\n",
    "\n",
    "To download data from SRA you will need to use the SRA-toolkit \n",
    "\n",
    "You can download reads from ENA using `wget` by \n",
    "- searching for the Accession in the box labeled ‘Enter accession’ \n",
    "- Select the check boxes in the ‘FASTQ FTP’ column for the samples you want and then download the report in .TSV format (click the ‘TSV’ link).\n",
    "- The 7th column (‘sra_ftp’) will have the FTP address(es) you can use to download the reads using wget, which will download the reads in your current working directory:\n",
    " \n",
    " `wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR142/025/SRR14227125/SRR14227125_1.fastq.gz`\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a5598c7",
   "metadata": {},
   "source": [
    "# Alignment formats\n",
    "These are formats for storing alignments of nucleotide or amino acid sequences. The alignment formats discussed here can be manipulated using SAMtools.\n",
    "\n",
    "## SAM\n",
    "Sequence Alignment/Map (SAM) is a text-based alignment format that supports single- and paired-end reads produced by different sequencing platforms. It can support short and long reads (up to 128Mbp). The format has been extended to include unmapped sequences, and it may contain other data such as base-call & alignment qualities. (https://samtools.github.io/hts-specs/SAMv1.pdf)\n",
    "\n",
    "The SAM format consists of a header and an alignment section. Headings begin with the @ symbol, which distinguishes them from the alignment section. All lines in a SAM file are tab-delimited. Alignment sections contain 11 mandatory fields, with other fields being optional. Although the mandatory fields must be present, their value can be a * or a 0 depending on the field. Optional fields are presented as key-value pairs in the format TAG:TYPE:VALUE."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cbc2b2d",
   "metadata": {},
   "source": [
    "### SAM HEADER\n",
    "\n",
    "Each line in the header section starts with an `@` symbol and a two-letter record type code.\n",
    "`@HD` line contains metadata and must be single first line if present \n",
    "`@PG` contains program-level information\n",
    "`@SQ` contains information about the reference sequence\n",
    "`@RG` is contains information about the different read groups (absent in the following example).\n",
    "\n",
    "Example header of a SAM file\n",
    "\n",
    "```\n",
    "@HD    VN:1.0    SO:unsorted\n",
    "@PG    ID:GSNAP    PN:gsnap    VN:2017-01-14    CL:/gpfs/runtime/cbc_conda/cbc_conda_v1_root/envs/cbc_conda_v1/bin/gsnap.nosimd -t 8 --gunzip -A sam -N1 --use-shared-memory=0 -d Ensembl_mus_GRCm38.p5_rel89 -s /gpfs/data/cbc/cbcollab/cbc_ref/gmapdb_2017.01.14/Ensembl_mus_GRCm38.p5_rel89/Ensembl_mus_GRCm38.p5_rel89.maps/Mus_musculus.GRCm38.89.splicesites.iit /gpfs/data/cbc/Muscle_RNA-seq_Jaime/fastq/SRS643406_SRX612440.fq.gz\n",
    "@SQ    SN:1    LN:195471971\n",
    "@SQ    SN:2    LN:182113224\n",
    "@SQ    SN:3    LN:160039680\n",
    "@SQ    SN:4    LN:156508116\n",
    "@SQ    SN:5    LN:151834684\n",
    "@SQ    SN:6    LN:149736546\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20e2e727-5fbb-4997-80e3-dc4450a94260",
   "metadata": {},
   "source": [
    "### Alignment Section\n",
    "\n",
    "The mandatory fields in a SAM file are…\n",
    "\n",
    "- QNAME Query template name. Reads/segments with the same QNAME are from the same template. A read may occupy multiple alignment lines when its alignment is chimeric or when multiple mappings are given.\n",
    "- FLAG Bitwise flag (pairing, strand, mate strand, etc.).\n",
    "- RNAME Reference sequence name.\n",
    "- POS 1-based leftmost mapping position. The first base in a reference sequence has coordinate 1. POS is set to 0 for an unmapped read.\n",
    "- MAPQ Mapping quality. If equal to 255, the mapping quality is not available.\n",
    "- CIGAR Extended Consise Idiosyncratic Gapped Alignment Report string. M for match/mismatch, I for insertion and D for deletion compared with the reference, N for skipped bases on the reference, S for soft clipping, H for hard clipping, and P for padding.\n",
    "- RNEXT Reference name of the mate/next read in the template. For the last read, the next read is the first read in the template.\n",
    "- PNEXT Position of the primary alignment of the mate/next read.\n",
    "- TLEN Observed template length.\n",
    "- SEQ Segment sequence.\n",
    "- QUAL Phred-based base quality (same as the quailty string in FASTQ) plus 33.\n",
    "\n",
    "A detailed specification is provided here https://samtools.github.io/hts-specs/SAMv1.pdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cfa5c84",
   "metadata": {},
   "source": [
    "Example Sam Alignment:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45297b9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "GWNJ-0957:353:GW1901111756:5:1208:18944:28927\t99\t1\t11644\t40\t85M3459N65M\t=\t15237\t284\tGCAGAAGGCTGCGGAGCGCCAGGACTTCGGACGCCCTCTGGCCTCGCGTCCTGGCTGGCCCTAGGTACAGCTCGGCCAGCCCACGAGATTAGAAGTGGCTGCTGAACTGTTTTCCAAAGGATAAACATTACAAATCATTGACTATGGCAG\tAAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ<JJJJJJJAJJJJJJJJJAJJJJFFFFJFJJJJJJJJJJJJJJJJ<JFJJJJJJJJJJJJJJJFJJJJFJFFA<FFJ<AMD:Z:150\tNH:i:1\tHI:i:1\tNM:i:0\tSM:i:40\tXQ:i:40\tX2:i:0\tXO:Z:CU\tXS:A:+\tXG:Z:A\n",
    "GWNJ-0957:353:GW1901111756:5:1201:10460:6425\t163\t1\t11681\t40\t48M3459N70M\t=\t20494\t5504\tCTGGCCTCGCGTCCTGGCTGGCCCTAGGTACAGCTCGGCCAGCCCACGAGAGTAGAAGTGGCTGCTTAACTGTTTTCCAAAGGATAAACATCACAAATCATTGACTAGGGCAGCTCAC\t-AAFF-FJJJJJJJJJJJJJJ-7F-FJJJJJJJF<JJJJJJFJJFJJJ7AJAJJ<FJJFJJJJJJ-<FJJ<AFF77FJJ7AAAJJJFJJA<-AFFJJFFJJJJFJ<F-AFF<A<-A<<\tMD:Z:51T14G24T15T10\tNH:i:1\tHI:i:1\tNM:i:4\tSM:i:40\tXQ:i:40\tX2:i:0\tXO:Z:CU\tXS:A:+\n",
    "GWNJ-0957:353:GW1901111756:5:1208:18944:28927\t147\t1\t15237\t40\t45M5171N105M\t=\t11644\t-284\tTCATTGACTATGGCAGCTCACTAAGAAGACCCTATTGCTTTGTGGGGACCACAACTTTCAGAAGCAGCCACAAGGGATGCTGTTACGCAGGTGCCAACCAAAACAAAGCCCACCTCGGGAGGTTTTCTGTGGGCTCTGCACCAGGATGGG\tJJJJJJJFJFJJJJJJJFJJJJJJJJJJJFFJJJJJJJJJFJJJJJJJJJJFJJJJJJJJJJJJFJAJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJAJJJJJJJJ<JJJJFJJJ<JJJJJJJJJJJJJAFJJJFFFAAMD:Z:42C10G15G60A19\tNH:i:1\tHI:i:1\tNM:i:4\tSM:i:40\tXQ:i:40\tX2:i:0\tXO:Z:CU\tXS:A:+\tXG:Z:A\n",
    "GWNJ-0957:353:GW1901111756:6:2122:8298:44767\t163\t1\t15259\t40\t23M4969N75M127N52M\t=\t20251\t171\tAAGAAGACCCTATTGCTTTGCGGATTTATTTTTTAAAGGAATGAATTGAGAGAAAAAACATGGGGAAGACAAGTATGGAACAGAAAATAAATATAAATGGACCACAGCTTTCAGAAGCAGCCGCAAGGGATGCTGTTACGCAGGTGCCAA\tAAAFFJJJJJJJJJJJFFJJJJFFJJJJJFJJJJJJFJJJFJJJJJFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJAJJFJJJJJAJJJAJFJJJJJJJJJJAFJJ<JFFJJJJJFFJF<JJJJJ-FJJ<AJFJJFJJJJJFJFAFJF\tMD:Z:150\tNH:i:1\tHI:i:1\tNM:i:0\tSM:i:40\tXQ:i:40\tX2:i:0\tXO:Z:CU\tXS:A:+\tXG:Z:A"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b67027b7",
   "metadata": {},
   "source": [
    "\n",
    "Understanding tags\n",
    "\n",
    "https://samtools.github.io/hts-specs/SAMtags.pdf\n",
    "\n",
    "\n",
    "Understanding SAM flags\n",
    "https://broadinstitute.github.io/picard/explain-flags.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de9245a3-c194-45eb-8ecd-45df4e9e4fd3",
   "metadata": {},
   "source": [
    "## BAM and CRAM\n",
    "\n",
    "SAM files can get very large but they can be converted into BAM files, which are just block compressed binary versions of SAM files that take up less space and allows quick access to a given alignment.\n",
    "- Like SAM files, they contain a header section and an alignment section, you can view the contents of a BAM file using `samtools view` (http://www.htslib.org/doc/samtools-view.html). \n",
    "- BAM files can be quickly queried if they are also indexed, which organizes your alignments into bins and is like making a table of contents for your alignments.\n",
    "- BAM indexes are stored in BAI files, use `samtools index` for this (http://www.htslib.org/doc/samtools-index.html).\n",
    "- A BAM file must be sorted before indexing, use `samtools sort` for this (http://www.htslib.org/doc/samtools-sort.html).\n",
    "- By default, `samtools sort` will sort based on chromosomal coordinates, but the `n` flag can be specified if read name sorting is required."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d874be2-0a68-45f6-bf9a-f006560e1dc1",
   "metadata": {},
   "source": [
    "# BED format\n",
    "\n",
    "BED files are tab or space-delimited text files to store genomic regions and annotations and are mainly used as a way to define which lines are shown in on an annotation track (can use with UCSC genome browser or IGV).\n",
    "\n",
    "There are only three required columns:\n",
    "- chrom: chromosome or scaffold name\n",
    "- chromStart: start site on chromosome (starts at 0)\n",
    "- chromEnd: end site on chromosome (up to this site will be shown -- chromStart=0 and chromEnd=100 will span bases 0-99 (but not 100). \n",
    "\n",
    "There’s also 9 additional optional fields:\n",
    "- name: name of the BED line.\n",
    "- score: A score between 0 and 1000 (UCSC browser will shade higher colors darker shades of grey).\n",
    "- strand: Strand orientation (-, +, or . if null/no strand)\n",
    "- thickStart: start coordinate where the annotation will be displayed in a thicker way (e.g., start codon)\n",
    "- thickEnd: stop coordinate where the annotation will stop being displayed in a thicker way (e.g., stop codon).\n",
    "- itemRgb: (R,G,B) color values determining the color that should be displayed for the annotation.\n",
    "- blockCount: number of blocks (exons) on the line of the BED\n",
    "- blockSizes: comma-separated values for the sizes of the blocks in blockCount\n",
    "- blockStarts: comma-separated values for the coordinates where the blocks start.\n",
    "\n",
    "\n",
    "Can be used for manipulating and the so called 'genome arithmetic' using the powerful bedtools suite\n",
    "https://bedtools.readthedocs.io/en/latest/\n",
    "\n",
    "Let's look at an example file first:\n",
    "\n",
    "```\n",
    "chr7  127471196  127472363  Pos1  0  +  127471196  127472363  255,0,0\n",
    "chr7  127472363  127473530  Pos2  0  +  127472363  127473530  255,0,0\n",
    "chr7  127473530  127474697  Pos3  0  +  127473530  127474697  255,0,0\n",
    "chr7  127474697  127475864  Pos4  0  +  127474697  127475864  255,0,0\n",
    "chr7  127475864  127477031  Neg1  0  -  127475864  127477031  0,0,255\n",
    "chr7  127477031  127478198  Neg2  0  -  127477031  127478198  0,0,255\n",
    "chr7  127478198  127479365  Neg3  0  -  127478198  127479365  0,0,255\n",
    "chr7  127479365  127480532  Pos5  0  +  127479365  127480532  255,0,0\n",
    "chr7  127480532  127481699  Neg4  0  -  127480532  127481699  0,0,255\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f6dba35-dc48-48e6-8645-600f6772e3e7",
   "metadata": {},
   "source": [
    "# GTF and GFF3 \n",
    "\n",
    "Reference genomes will have annotations associated with them, usually `.gtf` or `.gff3`, which are text files that describe different genomic features. \n",
    "Each of these files contain 9 tab-separated fields, none of which can be empty (but could have a `.` to indicate a null value instead of an empty field): \n",
    "\n",
    "- seqname - name of the chromosome or scaffold; chromosome names can be given with or without the 'chr' prefix.\n",
    "- source - name of the program that generated this feature, or the data source (database or project name)\n",
    "- feature - feature type, must be either \n",
    "- start - Start position of the feature, with sequence numbering starting at 1.\n",
    "- end - End position of the feature, with sequence numbering starting at 1.\n",
    "- score - A floating point value.\n",
    "- strand - defined as + (forward) or - (reverse).\n",
    "- frame or phase - One of '0', '1' or '2'. '0' indicates that the first base of the feature is the first base of a codon, '1' that the second base is the first base of a codon, and so on..\n",
    "- attribute - A semicolon-separated list of tag-value pairs, providing additional information about each feature.\n",
    "\n",
    "The main distinction is that `.gff3` files allow for 3 levels of hierarchical groupings of features and sub-features (i.e., genes, transcripts, exons), and allows each sub-feature to belong to more than one feature (i.e., an exon can be used in more than one gene or transcript).\n",
    "\n",
    "Note that .gff files are not the same as .gff3 files.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e281bff",
   "metadata": {},
   "source": [
    "Example of an GFF3 file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "612437c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ctg123 example gene           1050 9000 . + . ID=EDEN;Name=EDEN;Note=protein kinase\n",
    "\n",
    "ctg123 example mRNA           1050 9000 . + . ID=EDEN.1;Parent=EDEN;Name=EDEN.1;Index=1\n",
    "\n",
    "ctg123 example five_prime_UTR 1050 1200 . + . Parent=EDEN.1\n",
    "\n",
    "ctg123 example CDS             1201 1500 . + 0 Parent=EDEN.1\n",
    "\n",
    "ctg123 example CDS             3000 3902 . + 0 Parent=EDEN.1\n",
    "\n",
    "ctg123 example CDS             5000 5500 . + 0 Parent=EDEN.1\n",
    "\n",
    "ctg123 example CDS             7000 7608 . + 0 Parent=EDEN.1\n",
    "\n",
    "ctg123 example three_prime_UTR 7609 9000 . + . Parent=EDEN.1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "ctg123 example mRNA           1050 9000 . + . ID=EDEN.2;Parent=EDEN;Name=EDEN.2;Index=1\n",
    "\n",
    "ctg123 example five_prime_UTR 1050 1200 . + . Parent=EDEN.2\n",
    "\n",
    "ctg123 example CDS             1201 1500 . + 0 Parent=EDEN.2\n",
    "\n",
    "ctg123 example CDS             5000 5500 . + 0 Parent=EDEN.2\n",
    "\n",
    "ctg123 example CDS             7000 7608 . + 0 Parent=EDEN.2\n",
    "\n",
    "ctg123 example three_prime_UTR 7609 9000 . + . Parent=EDEN.2\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "ctg123 example mRNA           1300 9000 . + . ID=EDEN.3;Parent=EDEN;Name=EDEN.3;Index=1\n",
    "\n",
    "ctg123 example five_prime_UTR 1300 1500 . + . Parent=EDEN.3\n",
    "\n",
    "ctg123 example five_prime_UTR 3000 3300 . + . Parent=EDEN.3\n",
    "\n",
    "ctg123 example CDS             3301 3902 . + 0 Parent=EDEN.3\n",
    "\n",
    "ctg123 example CDS             5000 5500 . + 1 Parent=EDEN.3\n",
    "\n",
    "ctg123 example CDS             7000 7600 . + 1 Parent=EDEN.3\n",
    "\n",
    "ctg123 example three_prime_UTR 7601 9000 . + . Parent=EDEN.3\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6d84ef5",
   "metadata": {},
   "source": [
    "### Example GTF format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a4644bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "```\n",
    "chr1\tprocessed_transcript\texon\t11869\t12227\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000456328\"; exon_number \"1\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-002\"; exon_id \"ENSE00002234944\";\n",
    "\n",
    "chr1\tprocessed_transcript\texon\t12613\t12721\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000456328\"; exon_number \"2\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-002\"; exon_id \"ENSE00003582793\";\n",
    "\n",
    "chr1\tprocessed_transcript\texon\t13221\t14409\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000456328\"; exon_number \"3\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-002\"; exon_id \"ENSE00002312635\";\n",
    "\n",
    "===========================================\n",
    "\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t11872\t12227\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000515242\"; exon_number \"1\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-201\"; exon_id \"ENSE00002234632\";\n",
    "\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t12613\t12721\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000515242\"; exon_number \"2\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-201\"; exon_id \"ENSE00003608237\";\n",
    "\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t13225\t14412\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000515242\"; exon_number \"3\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-201\"; exon_id \"ENSE00002306041\";\n",
    "\n",
    "============================================\n",
    "\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t11874\t12227\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000518655\"; exon_number \"1\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-202\"; exon_id \"ENSE00002269724\";\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t12595\t12721\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000518655\"; exon_number \"2\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-202\"; exon_id \"ENSE00002270865\";\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t13403\t13655\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000518655\"; exon_number \"3\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-202\"; exon_id \"ENSE00002216795\";\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t13661\t14409\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000518655\"; exon_number \"4\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-202\"; exon_id \"ENSE00002303382\";\n",
    "\n",
    "=======================================\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t12010\t12057\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000450305\"; exon_number \"1\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-001\"; exon_id \"ENSE00001948541\";\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t12179\t12227\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000450305\"; exon_number \"2\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-001\"; exon_id \"ENSE00001671638\";\n",
    "\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t12613\t12697\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000450305\"; exon_number \"3\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-001\"; exon_id \"ENSE00001758273\";\n",
    "\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t12975\t13052\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000450305\"; exon_number \"4\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-001\"; exon_id \"ENSE00001799933\";\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t13221\t13374\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000450305\"; exon_number \"5\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-001\"; exon_id \"ENSE00001746346\";\n",
    "chr1\ttranscribed_unprocessed_pseudogene\texon\t13453\t13670\t.\t+\t.\tgene_id \"ENSG00000223972\"; transcript_id \"ENST00000450305\"; exon_number \"6\"; gene_name \"DDX11L1\"; gene_biotype \"pseudogene\"; transcript_name \"DDX11L1-001\"; exon_id \"ENSE00001863096\";\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d922d244",
   "metadata": {},
   "source": [
    "# VCF files\n",
    "\n",
    "VCF, or Variant Call Format, It is a standardized text file format used for representing SNP, indel, and structural variation calls. \n",
    "\n",
    "### Structure of a VCF file\n",
    "A valid VCF file is composed of two main parts: the header, and the variant call records.\n",
    "\n",
    "![vcf file](imgs/vcf_structure.png)\n",
    "\n",
    "The header contains information about the dataset and relevant reference sources (e.g. the organism, genome build version etc.), as well as definitions of all the annotations used to qualify and quantify the properties of the variant calls contained in the VCF file. The header of VCFs generated by GATK tools also include the command line that was used to generate them. Some other programs also record the command line in the VCF header, but not all do so as it is not required by the VCF specification. For more information about the header, see the next section.\n",
    "\n",
    "The actual data lines will look something like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6184c557",
   "metadata": {},
   "outputs": [],
   "source": [
    "[HEADER LINES]\n",
    "#CHROM  POS ID      REF ALT QUAL    FILTER  INFO          FORMAT          NA12878\n",
    "20  10001019    .   T   G   364.77  .   AC=1;AF=0.500;AN=2;BaseQRankSum=0.699;ClippingRankSum=0.00;DP=34;ExcessHet=3.0103;FS=3.064;MLEAC=1;MLEAF=0.500;MQ=42.48;MQRankSum=-3.219e+00;QD=11.05;ReadPosRankSum=-6.450e-01;SOR=0.537   GT:AD:DP:GQ:PL  0/1:18,15:33:99:393,0,480\n",
    "20  10001298    .   T   A   884.77  .   AC=2;AF=1.00;AN=2;DP=30;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=29.49;SOR=1.765    GT:AD:DP:GQ:PL  1/1:0,30:30:89:913,89,0\n",
    "20  10001436    .   A   AAGGCT  1222.73 .   AC=2;AF=1.00;AN=2;DP=29;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=25.36;SOR=0.836    GT:AD:DP:GQ:PL  1/1:0,28:28:84:1260,84,0\n",
    "20  10001474    .   C   T   843.77  .   AC=2;AF=1.00;AN=2;DP=27;ExcessHet=3.0103;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=31.25;SOR=1.302    GT:AD:DP:GQ:PL  1/1:0,27:27:81:872,81,0\n",
    "20  10001617    .   C   A   493.77  .   AC=1;AF=0.500;AN=2;BaseQRankSum=1.63;ClippingRankSum=0.00;DP=38;ExcessHet=3.0103;FS=1.323;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.00;QD=12.99;ReadPosRankSum=0.170;SOR=1.179   GT:AD:DP:GQ:PL  0/1:19,19:38:99:522,0,480"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c905b5c2",
   "metadata": {},
   "source": [
    "The VCF Header contains a ton of information that should be reviewed carefully"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f528ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "##fileformat=VCFv4.2\n",
    "##ALT=<ID=NON_REF,Description=\"Represents any possible alternative allele at this location\">\n",
    "##FILTER=<ID=LowQual,Description=\"Low quality\">\n",
    "##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">\n",
    "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">\n",
    "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\n",
    "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
    "##FORMAT=<ID=MIN_DP,Number=1,Type=Integer,Description=\"Minimum DP observed within the GVCF block\">\n",
    "##FORMAT=<ID=PGT,Number=1,Type=String,Description=\"Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another\">\n",
    "##FORMAT=<ID=PID,Number=1,Type=String,Description=\"Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group\">\n",
    "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
    "##FORMAT=<ID=RGQ,Number=1,Type=Integer,Description=\"Unconditional reference genotype confidence, encoded as a phred quality -10*log10 p(genotype call is wrong)\">\n",
    "##FORMAT=<ID=SB,Number=4,Type=Integer,Description=\"Per-sample component statistics which comprise the Fisher's Exact Test to detect strand bias.\">\n",
    "##GATKCommandLine.HaplotypeCaller=<ID=HaplotypeCaller,Version=3.7-0-gcfedb67,Date=\"Fri Jan 20 11:14:15 EST 2017\",Epoch=1484928855435,CommandLineOptions=\"[command-line goes here]\">\n",
    "##GATKCommandLine=<ID=GenotypeGVCFs,CommandLine=\"[command-line goes here]\",Version=4.beta.6-117-g4588584-SNAPSHOT,Date=\"December 23, 2017 5:45:56 PM EST\">\n",
    "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">\n",
    "##INFO=<ID=AF,Number=A,Type=Float,Description=\"Allele Frequency, for each ALT allele, in the same order as listed\">\n",
    "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
    "##INFO=<ID=BaseQRankSum,Number=1,Type=Float,Description=\"Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities\">\n",
    "##INFO=<ID=ClippingRankSum,Number=1,Type=Float,Description=\"Z-score From Wilcoxon rank sum test of Alt vs. Ref number of hard clipped bases\">\n",
    "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth; some reads may have been filtered\">\n",
    "##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">\n",
    "##INFO=<ID=END,Number=1,Type=Integer,Description=\"Stop position of the interval\">\n",
    "##INFO=<ID=ExcessHet,Number=1,Type=Float,Description=\"Phred-scaled p-value for exact test of excess heterozygosity\">\n",
    "##INFO=<ID=FS,Number=1,Type=Float,Description=\"Phred-scaled p-value using Fisher's exact test to detect strand bias\">\n",
    "##INFO=<ID=HaplotypeScore,Number=1,Type=Float,Description=\"Consistency of the site with at most two segregating haplotypes\">\n",
    "##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">\n",
    "##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">\n",
    "##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">\n",
    "##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">\n",
    "##INFO=<ID=MQRankSum,Number=1,Type=Float,Description=\"Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities\">\n",
    "##INFO=<ID=QD,Number=1,Type=Float,Description=\"Variant Confidence/Quality by Depth\">\n",
    "##INFO=<ID=RAW_MQ,Number=1,Type=Float,Description=\"Raw data for RMS Mapping Quality\">\n",
    "##INFO=<ID=ReadPosRankSum,Number=1,Type=Float,Description=\"Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias\">\n",
    "##INFO=<ID=SOR,Number=1,Type=Float,Description=\"Symmetric Odds Ratio of 2x2 contingency table to detect strand bias\">\n",
    "##contig=<ID=20,length=63025520>\n",
    "##reference=file:///data/ref/ref.fasta\n",
    "##source=GenotypeGVCFs"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
